app/javascript/components/student/Editor.tsx

```diff
@@ -0,0 +1,45 @@
+import React, { useState } from 'react'
+import { CodeEditor } from './editor/CodeEditor'
+import { TestRunSummary } from './editor/TestRunSummary'
+import fetch from 'isomorphic-fetch'
+
+export type Submission = {
+  testsStatus: TestRunStatus
+  uuid: string
+}
+
+export enum TestRunStatus {
+  PASS = 'pass',
+  FAIL = 'fail',
+  ERROR = 'error',
+  PENDING = 'pending',
+  OPS_ERROR = 'ops_error',
+}
+
+export function Editor({ endpoint }: { endpoint: string }) {
+  const [submission, setSubmission] = useState<Submission>()
+
+  function submit(code: string) {
+    setSubmission(undefined)
+
+    fetch(endpoint, {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify({ files: { file: code } }),
+    })
+      .then((response) => response.json())
+      .then(({ submission }: any) => {
+        setSubmission({
```

Comment:
There are a few issues with this:

1. By the time you get here, the component may be unmounted and thus this will lead to a "unmounted component tried to render" error.

You can use my npm package `use-is-mounted` or roll your own:

```typescript
function Component() {
  const isMountedRef = useIsMounted()

  // ...

  .then((...) => {
    if (!isMountedRef.current) {
      return
    }

    setSubmission(...)
  })
```

2. Race condition! You must check inside the then if the submission hasn't changed in-between. For example you could check if the code has changed.

Far easier is to keep track with a nonce/state/timestamp/whatever in a custom ref, and check if it's still the same as before you started fetch. If so, yay. Otherwise ignore.

3. Both issues above can (partially) be mitigated by cancelling the request when the component unmounts or a new request comes in. Do this by using an abortcontroller, and adding a signal to the `fetch`. When a new submissions request is made, before the fetch, cancel the existing abortcontroller. Do the same on unmount:

```typescript
useEffect(() => () => {
  // Check if there is an existing abortcontroller and cancel it
}, [])
```

Reply:
I went with 3 for now.

Reply:
(These were not options, but 3 issues!)